framework/include/gpu/base/GPUArray.h

```diff
@@ -0,0 +1,1529 @@
+
+  _device_data = ptr;
+  _is_device_alloc = true;
+  _is_device_alias = true;
+}
+
+template <typename T, unsigned int dimension>
+void
+ArrayBase<T, dimension>::allocHost()
+{
+  if (_is_host_alloc)
+    return;
+
+  if constexpr (std::is_default_constructible<T>::value)
+    _host_data = new T[_size];
+  else
+    _host_data = static_cast<T *>(std::malloc(_size * sizeof(T)));
+
+  _is_host_alloc = true;
+}
+
+template <typename T, unsigned int dimension>
+void
+ArrayBase<T, dimension>::allocDevice()
+{
+  if (_is_device_alloc)
+    return;
+
+  _device_data =
+      static_cast<T *>(::Kokkos::kokkos_malloc<ExecSpace::memory_space>(_size * sizeof(T)));
+
+  _is_device_alloc = true;
+}
+
+template <typename T, unsigned int dimension>
+template <bool host, bool device>
+void
+ArrayBase<T, dimension>::createInternal(const std::vector<dof_id_type> n)
+{
+  if (n.size() != dimension)
+    mooseError("Kokkos array error: the number of dimensions provided (",
+               n.size(),
+               ") must match the array dimension (",
+               dimension,
+               ").");
+
+  if (_counter)
+    destroy();
+
+  _counter = new unsigned int;
+  *_counter = 1;
+
+  _size = 1;
+  _s[0] = 1;
+
+  for (unsigned int i = 0; i < dimension; ++i)
+  {
+    _n[i] = n[i];
+    _size *= n[i];
+
+    if (i)
+      _s[i] = _s[i - 1] * _n[i - 1];
+  }
+
+  if constexpr (host)
+    allocHost();
+
+  if constexpr (device)
+    allocDevice();
+}
+
+template <typename T, unsigned int dimension>
+void
+ArrayBase<T, dimension>::createInternal(const std::vector<dof_id_type> n, bool host, bool device)
+{
+  if (host && device)
+    createInternal<true, true>(n);
+  else if (host && !device)
+    createInternal<true, false>(n);
+  else if (!host && device)
+    createInternal<false, true>(n);
+  else
+    createInternal<false, false>(n);
+}
+
+template <typename T, unsigned int dimension>
+template <typename TargetSpace, typename SourceSpace>
+void
+ArrayBase<T, dimension>::copyInternal(T * target, const T * source, dof_id_type n)
+{
+  ::Kokkos::Impl::DeepCopy<TargetSpace, SourceSpace>(target, source, n * sizeof(T));
+  ::Kokkos::fence();
+}
+
+template <typename T, unsigned int dimension>
+void
+ArrayBase<T, dimension>::offset(const std::vector<dof_id_signed_type> d)
+{
+  if (d.size() > dimension)
+    mooseError("Kokkos array error: the number of offsets provided (",
+               d.size(),
+               ") cannot be larger than the array dimension (",
+               dimension,
+               ").");
+
+  for (unsigned int i = 0; i < d.size(); ++i)
+    _d[i] = d[i];
+}
+
+template <typename T, unsigned int dimension>
+void
+ArrayBase<T, dimension>::copy(MemcpyKind dir)
+{
+  if (dir == MemcpyKind::HOST_TO_DEVICE)
+  {
+    // If host side memory is not allocated, do nothing
+    if (!_is_host_alloc)
+      return;
+
+    // If device side memory is not allocated,
+    if (!_is_device_alloc)
+    {
+      if (*_counter == 1)
```

Comment:
The aliasing stuff makes me nervous because I've already had the concept bite me in MFEM. How often are we aliasing? Is there any reason to consider reusing code in MFEM? This class in particular has many parallels to https://github.com/mfem/mfem/blob/master/general/mem_manager.hpp. I'm guessing it would be impossible to use that given the underlying Kokkos use here